File: frontend/src/features/analysis/policy.ts
import type { AnalyzedComment, ModerationAction, SecurityLevel } from '../../api/types';

// 백엔드 PolicyManager와 동일한 매트릭스 (policy_manager.py _MATRIX 미러).
// 보안 모드 변경 시 서버 재호출 없이 action/displayText를 재유도하기 위해 미러링.
type Category = 'BLACKLIST' | 'GENERAL_TRIGGER' | 'GENERAL';

const MATRIX: Record<SecurityLevel, Record<Category, ModerationAction>> = {
  LOW: {
    BLACKLIST: 'PARTIAL_MASK',
    GENERAL_TRIGGER: 'PARTIAL_MASK',
    GENERAL: 'REVIEW',
  },
  MEDIUM: {
    BLACKLIST: 'FULL_BLOCK',
    GENERAL_TRIGGER: 'PARTIAL_MASK',
    GENERAL: 'PARTIAL_MASK',
  },
  HIGH: {
    BLACKLIST: 'FULL_BLOCK',
    GENERAL_TRIGGER: 'FULL_BLOCK',
    GENERAL: 'PARTIAL_MASK',
  },
};

const AI_MATRIX: Record<SecurityLevel, ModerationAction> = {
  LOW: 'REVIEW',
  MEDIUM: 'FULL_BLOCK',
  HIGH: 'FULL_BLOCK',
};

export interface DerivedDecision {
  action: ModerationAction;
  processed_text: string;
  score: number;
}

// 서버 base score 에 보안수준 가산(+25)을 적용한 display score.
// Caution 버킷(SYSTEM 전용): MEDIUM·HIGH 에서 +25 (MEDIUM 도 이제 PARTIAL_MASK 로 처분됨)
// Warn 버킷(BLACKLIST): MEDIUM·HIGH 에서 +25
// 보안수준 변경 시 서버 재호출 없이 즉시 반응하도록 클라이언트에서 계산한다.
const applySecurityAdjustment = (
  baseScore: number,
  comment: AnalyzedComment,
  securityLevel: SecurityLevel,
): number => {
  const hasAiModules = (comment.ai_modules ?? []).length > 0;

  if (!comment.detected_words.length && !hasAiModules) {
    return baseScore;
  }

  let bonus = 0;

  if (hasAiModules) {
    if (securityLevel === 'HIGH') bonus = 0.25;
  } else if (comment.flags.has_blacklist) {
    if (securityLevel === 'MEDIUM' || securityLevel === 'HIGH') bonus = 0.25;
  } else if (comment.flags.has_general) {
    if (securityLevel === 'MEDIUM' || securityLevel === 'HIGH') bonus = 0.25;
  }

  return Math.min(1.0, baseScore + bonus);
};

export const derivePolicy = (
  comment: AnalyzedComment,
  securityLevel: SecurityLevel,
): DerivedDecision => {
  const score = applySecurityAdjustment(comment.score, comment, securityLevel);
  const hasAiModules = (comment.ai_modules ?? []).length > 0;

  // 2차 AI 탐지 결과
  // AI는 특정 단어 위치를 모르므로 PARTIAL_MASK를 사용하지 않는다.
  if (hasAiModules) {
    const action = AI_MATRIX[securityLevel];

    return {
      action,
      processed_text: action === 'FULL_BLOCK' ? '' : comment.text,
      score,
    };
  }
  
  if (!comment.detected_words.length) {
    return { action: 'NORMAL', processed_text: comment.text, score };
  }

  // detected_words 가 존재하면 어느 카테고리로든 분류되어야 한다.
  // 플래그가 모두 false 로 돌아오는 엣지케이스(신규 WordType 등)는 GENERAL 로 폴백.
  const category: Category = comment.flags.has_blacklist
    ? 'BLACKLIST'
    : comment.flags.has_trigger
    ? 'GENERAL_TRIGGER'
    : 'GENERAL';

  const action = MATRIX[securityLevel][category];
  const processed_text =
    action === 'FULL_BLOCK' ? '' : action === 'PARTIAL_MASK' ? comment.masked_text : comment.text;

  return { action, processed_text, score };
};

File: packages/nerv-filter/src/nerv_filter/core/policy.py
"""정책 매니저 — scorer 결과 + 보안수준으로 최종 처분 결정."""
from __future__ import annotations

import logging
from typing import Any, Dict, List

from ..models import ModerationAction, SecurityLevel, WordType

logger = logging.getLogger(__name__)


_CAT_BLACKLIST = "BLACKLIST"
_CAT_GENERAL_TRIGGER = "GENERAL_TRIGGER"
_CAT_GENERAL = "GENERAL"


# 매트릭스: [보안모드][검출카테고리] → 액션
_MATRIX: Dict[SecurityLevel, Dict[str, ModerationAction]] = {
    SecurityLevel.LOW: {
        _CAT_BLACKLIST: ModerationAction.PARTIAL_MASK,
        _CAT_GENERAL_TRIGGER: ModerationAction.PARTIAL_MASK,
        _CAT_GENERAL: ModerationAction.REVIEW,
    },
    SecurityLevel.MEDIUM: {
        _CAT_BLACKLIST: ModerationAction.FULL_BLOCK,
        _CAT_GENERAL_TRIGGER: ModerationAction.PARTIAL_MASK,
        _CAT_GENERAL: ModerationAction.PARTIAL_MASK,
    },
    SecurityLevel.HIGH: {
        _CAT_BLACKLIST: ModerationAction.FULL_BLOCK,
        _CAT_GENERAL_TRIGGER: ModerationAction.FULL_BLOCK,
        _CAT_GENERAL: ModerationAction.PARTIAL_MASK,
    },
}


class PolicyManager:
    """검출 결과 + 보안수준 → ModerationAction.

    SDK 공개 API 는 아님. NervFilter 내부에서 사용.
    """

    def __init__(self) -> None:
        logger.debug("[nerv_filter] PolicyManager 로드")

    def decide_action(
        self,
        scorer_result: Dict[str, Any],
        filter_result: Dict[str, Any],
        security_level: SecurityLevel | str,
    ) -> Dict[str, Any]:
        original_text = filter_result.get("original_text", "")
        detected_words = filter_result.get("detected_words", [])
        score = scorer_result.get("score", 0.0)

        if not detected_words:
            return {
                "action": ModerationAction.NORMAL,
                "processed_text": original_text,
                "score": score,
            }

        if scorer_result.get("has_blacklist"):
            category = _CAT_BLACKLIST
        elif scorer_result.get("has_trigger"):
            category = _CAT_GENERAL_TRIGGER
        else:
            category = _CAT_GENERAL

        level = (
            security_level
            if isinstance(security_level, SecurityLevel)
            else SecurityLevel(security_level)
        )
        action = _MATRIX[level][category]

        processed_text = self._render_text(action, original_text, detected_words)

        return {
            "action": action,
            "processed_text": processed_text,
            "score": score,
        }

    def _render_text(
        self,
        action: ModerationAction,
        original_text: str,
        detected_words: List[Dict[str, Any]],
    ) -> str:
        if action == ModerationAction.FULL_BLOCK:
            return ""
        if action == ModerationAction.PARTIAL_MASK:
            return self._star_mask(original_text, detected_words)
        return original_text

    @staticmethod
    def _star_mask(text: str, detected_words: List[Dict[str, Any]]) -> str:
        masked = text
        for item in detected_words:
            word = item.get("word", "")
            if word and item.get("type") != WordType.WHITELIST:
                masked = masked.replace(word, "*" * len(word))
        return masked
